sql/src/test/java/io/crate/execution/engine/collect/collectors/OrderedLuceneBatchIteratorFactoryTest.java

```diff
@@ -178,6 +182,39 @@ public void testOrderedLuceneBatchIteratorWithMultipleCollectorsTripsCircuitBrea
         consumeIteratorAndVerifyResultIsException(rowBatchIterator, circuitBreakingException);
     }
 
+    @Test
+    public void test_lucene_ordered_collector_thread_dies_on_kill() throws Exception {
+        LuceneOrderedDocCollector luceneOrderedDocCollector = createOrderedCollector(searcher1, 1);
+        AtomicReference<Thread> collectThread = new AtomicReference<>();
+        CountDownLatch latch = new CountDownLatch(1);
+        BatchIterator<Row> rowBatchIterator = OrderedLuceneBatchIteratorFactory.newInstance(
+            Collections.singletonList(luceneOrderedDocCollector),
+            OrderingByPosition.rowOrdering(new int[]{0}, reverseFlags, nullsFirst),
+            mock(RowAccounting.class),
+            c -> {
+                var t = new Thread(c);
+                collectThread.set(t);
+                t.start();
+                latch.countDown();
+            },
+            () -> 1,
+            true
+        );
+        TestingRowConsumer consumer = new TestingRowConsumer();
+        consumer.accept(rowBatchIterator, null);
+
+        // Ensure that the collect thread is running
+        latch.await(1, TimeUnit.SECONDS);
+        assertThat(collectThread.get().isAlive(), is(true));
+
+        rowBatchIterator.kill(new InterruptedException("killed"));
+
+        // The collect thread must stop "almost" immediately, 10ms should be enough to wait.
+        // Without any cancel logic in-place the thread operation will continue and may not complete in 10ms.
+        assertBusy(() -> assertThat(collectThread.get().isAlive(), is(false)),
```

Comment:
I'm not soo happy with this assertion but I found no other way as the thread will always stop eventually.
Locally it always failed without the cancel logic in place. 
Strangely, checking for `isInterrupted` did not always succeed even with higher timeouts. Afaik it could be that the flag was reset while checking for interruption internally.

Reply:
How about something like:

```
--- a/sql/src/test/java/io/crate/execution/engine/collect/collectors/OrderedLuceneBatchIteratorFactoryTest.java
+++ b/sql/src/test/java/io/crate/execution/engine/collect/collectors/OrderedLuceneBatchIteratorFactoryTest.java
@@ -32,6 +32,7 @@ import io.crate.breaker.RowAccountingWithEstimators;
 import io.crate.data.BatchIterator;
 import io.crate.data.Row;
 import io.crate.exceptions.CircuitBreakingException;
+import io.crate.exceptions.SQLExceptions;
 import io.crate.execution.engine.sort.OrderingByPosition;
 import io.crate.expression.reference.doc.lucene.CollectorContext;
 import io.crate.expression.reference.doc.lucene.LuceneCollectorExpression;
@@ -58,6 +59,7 @@ import org.elasticsearch.common.UUIDs;
 import org.elasticsearch.common.breaker.CircuitBreaker;
 import org.elasticsearch.common.breaker.NoopCircuitBreaker;
 import org.elasticsearch.index.shard.ShardId;
+import org.hamcrest.Matchers;
 import org.junit.Before;
 import org.junit.Test;

@@ -209,10 +211,12 @@ public class OrderedLuceneBatchIteratorFactoryTest extends CrateUnitTest {

         rowBatchIterator.kill(new InterruptedException("killed"));

-        // The collect thread must stop "almost" immediately, 10ms should be enough to wait.
-        // Without any cancel logic in-place the thread operation will continue and may not complete in 10ms.
-        assertBusy(() -> assertThat(collectThread.get().isAlive(), is(false)),
-                   10, TimeUnit.MILLISECONDS);
+        try {
+            consumer.completionFuture().get(5, TimeUnit.SECONDS);
+        } catch (Exception e) {
+            var t = SQLExceptions.unwrap(e);
+            assertThat(t, Matchers.instanceOf(InterruptedException.class));
+        }
     }
```

Reply:
This would not validate the change (the test would succeed even without any change) as the exception was correctly set on the future inside the `BatchPageIterator` already.